geopandas/geodataframe.py

```diff
@@ -324,24 +338,20 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
                     "the column name is shared by multiple columns."
                 )
 
-            if drop:
-                to_remove = col
-            else:
-                geo_column_name = col
+            if drop and geo_column_name in frame.columns:
+                # if _geometry_column_name is None, geo_column_name
+                # may not exist
+                del frame[geo_column_name]
 
-        if to_remove:
-            del frame[to_remove]
+            geo_column_name = col
 
         if not crs:
             crs = getattr(level, "crs", None)
 
-        if isinstance(level, (GeoSeries, GeometryArray)) and level.crs != crs:
-            # Avoids caching issues/crs sharing issues
-            level = level.copy()
-            level.crs = crs
-
         # Check that we are using a listlike of geometries
         level = _ensure_geometry(level, crs=crs)
+        # ensure_geometry only sets crs on level if it has crs==None
+        level.crs = crs
```

Comment:
This double copy (there is a copy around crs caching inside `_ensure_geometry` too) doesn't seem to be needed.

Reply:
Good catch that `_ensure_geometry` indeed already has the same logic!